driver-examples/src/main/java/com/datastax/driver/examples/retry/DowngradingRetry.java

```diff
@@ -0,0 +1,467 @@
+     * Queries data, retrying if necessary with a downgraded CL.
+     *
+     * @param cl         the consistency level to apply.
+     * @param retryCount the current retry count.
+     * @throws DriverException if the current consistency level cannot be downgraded.
+     */
+    private ResultSet read(ConsistencyLevel cl, int retryCount) {
+
+        System.out.printf("Reading at %s (retry count: %d)%n", cl, retryCount);
+
+        Statement stmt = new SimpleStatement("SELECT sensor_id, date, timestamp, value " +
+                "FROM downgrading.sensor_data " +
+                "WHERE " +
+                "sensor_id = 756716f7-2e54-4715-9f00-91dcbea6cf50 AND " +
+                "date = '2018-02-26' AND " +
+                "timestamp > '2018-02-26+01:00'")
+                .setConsistencyLevel(cl);
+
+        try {
+
+            ResultSet rows = session.execute(stmt);
+            System.out.println("Read succeeded at " + cl);
+            return rows;
+
+        } catch (DriverException e) {
+
+            if (retryCount == maxRetries) {
+                throw e;
+            }
+
+            e = unwrap(e);
+
+            System.out.println("Read failed: " + e);
+
+            // General intent: downgrade and retry at the highest consistency level
+            // that is likely to succeed.
+
+            if (e instanceof UnavailableException) {
+
+                // Downgrade to the number of replicas reported alive and retry.
+                int aliveReplicas = ((UnavailableException) e).getAliveReplicas();
+
+                ConsistencyLevel downgraded = downgrade(cl, aliveReplicas, e);
+                return read(downgraded, retryCount + 1);
+
+            } else if (e instanceof ReadTimeoutException) {
+
+                ReadTimeoutException readTimeout = (ReadTimeoutException) e;
+                int received = readTimeout.getReceivedAcknowledgements();
+                int required = readTimeout.getRequiredAcknowledgements();
+
+                // If fewer replicas responded than required by the consistency level
+                // (but at least one replica did respond), retry with a consistency level
+                // equal to the number of received acknowledgements.
+                if (received < required) {
+
+                    ConsistencyLevel downgraded = downgrade(cl, received, e);
+                    return read(downgraded, retryCount + 1);
+
+                }
+
+                // If we received enough replies to meet the consistency level,
+                // but the actual data was not present among the received responses,
+                // then retry with the initial consistency level, we might be luckier next time
+                // and get the data back.
+                if (!readTimeout.wasDataRetrieved()) {
+
+                    return read(cl, retryCount + 1);
+
+                }
+
+                // Otherwise, abort since the read timeout is probably a more serious problem.
+                throw e;
```

Comment:
A question came to my mind while writing this example: in which circumstances can we get here? If we got the data _and_ enough acks, how can the read timeout?